src/celeritas/em/model/MuPairProductionModel.cc

```diff
@@ -0,0 +1,177 @@
+    host_data.ids.mu_plus = particles.find(pdg::mu_plus());
+    host_data.ids.electron = particles.find(pdg::electron());
+    host_data.ids.positron = particles.find(pdg::positron());
+    CELER_VALIDATE(host_data.ids,
+                   << "missing particles (required for '"
+                   << this->description() << "')");
+
+    // Save particle properties
+    host_data.electron_mass = particles.get(host_data.ids.electron).mass();
+
+    // Build sampling table
+    CELER_VALIDATE(imported,
+                   << "sampling table (required for '" << this->description()
+                   << "') is empty");
+    this->build_table(imported, &host_data.table);
+
+    // Move to mirrored data, copying to device
+    data_ = CollectionMirror<MuPairProductionData>{std::move(host_data)};
+
+    CELER_ENSURE(data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Particle types and energy ranges that this model applies to.
+ */
+auto MuPairProductionModel::applicability() const -> SetApplicability
+{
+    Applicability mu_minus_applic;
+    mu_minus_applic.particle = this->host_ref().ids.mu_minus;
+    mu_minus_applic.lower = zero_quantity();
+    mu_minus_applic.upper = detail::high_energy_limit();
+
+    Applicability mu_plus_applic = mu_minus_applic;
+    mu_plus_applic.particle = this->host_ref().ids.mu_plus;
+
+    return {mu_minus_applic, mu_plus_applic};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the microscopic cross sections for the given particle and material.
+ */
+auto MuPairProductionModel::micro_xs(Applicability applic) const
+    -> MicroXsBuilders
+{
+    return imported_.micro_xs(std::move(applic));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Interact with host data.
+ */
+void MuPairProductionModel::step(CoreParams const& params,
+                                 CoreStateHost& state) const
+{
+    auto execute = make_action_track_executor(
+        params.ptr<MemSpace::native>(),
+        state.ptr(),
+        this->action_id(),
+        InteractionApplier{MuPairProductionExecutor{this->host_ref()}});
+    return launch_action(*this, params, state, execute);
+}
+
+//---------------------------------------------------------------------------//
+#if !CELER_USE_DEVICE
+void MuPairProductionModel::step(CoreParams const&, CoreStateDevice&) const
+{
+    CELER_NOT_CONFIGURED("CUDA OR HIP");
+}
+#endif
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct sampling table.
+ */
+void MuPairProductionModel::build_table(
+    ImportMuPairProductionTable const& imported, HostTable* table) const
+{
+    CELER_EXPECT(imported);
+    CELER_EXPECT(table);
+
+    // Build 2D sampling table
+    TwodGridBuilder build_grid{&table->reals};
+    auto grids = make_builder(&table->grids);
```

Comment:
Since C++17 we've been able to do
```suggestion
    CollectionBuilder grids{&table->grids};
```